### PR TITLE
drivers: Optimized periph PWM interfaces

### PR DESCRIPTION
**Pull request by haukepetersen**
Updated the PWM interface slightly:
- changed some types to explicit width
- removed return value from `pwm_set`
- made types definable by CPUs
- EDIT: simplified error return code of `pwm_init`

I did however not yet optimize any of the existing PWM drivers with the new possibilities...

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
@A-Paul might be interested

File: drivers/include/periph/pwm.h [hunk 1]
```diff
@@ -18,84 +18,91 @@
...
+ * @{
  */
-typedef enum {
-#if PWM_0_EN
-    PWM_0,              /*< 1st PWM device */
-#endif
-#if PWM_1_EN
-    PWM_1,              /*< 2nd PWM device */
+#ifndef PWM_DEV
+#define PWM_DEV(x)          (x)
```
Comment (A-Paul) [inline drivers/include/periph/pwm.h]: I don't understand this.
Also found no place in your changes where it is used.

File: drivers/include/periph/pwm.h [hunk 2]
```diff
@@ -18,84 +18,91 @@
...
- * @param[in] frequency     the PWM frequency in Hz
- * @param[in] resolution    the PWM resolution
+ * @param[in] freq          PWM frequency in Hz
+ * @param[in] res           PWM resolution
  *
- * @return                  Actual PWM frequency on success
- * @return                  -1 on mode not applicable
- * @return                  -2 on frequency and resolution not applicable
+ * @return                  actual PWM frequency on success
+ * @return                  0 on error
```
Comment (A-Paul) [inline drivers/include/periph/pwm.h]: How do existing applications (their implementers) get aware that their error checking got invalid?

**Comment by A-Paul** [comment]
In your change list you forgot to mention the interface change for `pwm_init()` regarding error codes.

File: drivers/include/periph/pwm.h (hunk 1 above)
Comment (haukepetersen) [inline drivers/include/periph/pwm.h]: It is not used (yet). This define provides a universal means of accessing the different PWM devices in a more programmable fashion, while still providing the possibility for each CPU to override this default behavior. Have a look at the `GPIO` define and it's usage in `drivers/include/periph/gpio.h` and how some CPUs use it (e.g. `cpu/stm32f4/include/periph_cpu.h`)...

File: drivers/include/periph/pwm.h (hunk 2 above)
Comment (haukepetersen) [inline drivers/include/periph/pwm.h]: they do not (that is the thing about API changes...). But to my knowledge are (i) not very many users of this API really using these return values, and more importantly (ii) are most implementations not actually returning the correct values...

**Comment by haukepetersen** [comment]
@A-Paul: added it to the list of changes.

rebased.

**Comment by haukepetersen** [comment]
@A-Paul ping?!

**Comment by haukepetersen** [comment]
rebased

**Comment by A-Paul** [comment]
> @A-Paul ping?!

Huh??
Gimme a hint. What did I miss?

**Comment by LudwigKnuepfer** [comment]
@a-paul I think @haukepetersen wants your approval.

**Comment by haukepetersen** [comment]
Yap, the all reoccurring question about the ack :-)

But I just see, that Travis is complaining, I look into it quickly

**Comment by A-Paul** [comment]
I'm kind of sorry for the latency ...
However, I just wanted to assist by commenting some things that attracted my attention. I didn't realize that I'm suddenly in charge for the final ack, particularly without being assigned.
Furthermore I'm not shure if I have the needed insight to check all the consequences of that API change.

**Comment by haukepetersen** [comment]
No problem. Now you are assigned :-) But nevertheless, not being assigned doesn't mean you can't ack something...

I reworked this PR again slightly by adding a `pwm_channels()` function that is very useful for testing and run-time channel assignments. Also Travis should now hopefully be happy...

**Comment by PeterKietzmann** [comment]
@A-Paul I think you have the needed insights and also Travis will be your helping hand

**Comment by LudwigKnuepfer** [comment]
@A-Paul Don't Panic!

**Comment by haukepetersen** [comment]
Travis is not happy -> I will introduce a fix for that separately, so lets put this PR on hold until the fix is merged. I will update this PR then accordingly...

**Comment by authmillenon** [comment]
What's the fix? Which PR is it waiting on?

**Comment by haukepetersen** [comment]
#3420

**Comment by haukepetersen** [comment]
Ok, added some guards to the old style PWM driver implementations, now Travis should be happy with this PR. I would propose to merge this if anyone is ok with the API changes and than introduce updated PWM drivers for the affected CPUs separately...

So this PR is not waiting on any other PR anymore, all it needs is an ACK and squashing...

**Comment by DipSwitch** [comment]
The changes seem alright as quick as I can see. Here we can also remap the
hardware assigned channels to logic channels? The board I'm making for
example has TIM4-CH3 connected to the microbus PWM pin. So imho it would be
nice to map PWM line 0 to hardware channel 3.

Don't know if this is feasible or of this would have serious flash usage
penalties?

**Comment by haukepetersen** [comment]
Not sure I understand you problem correctly. Do you mean mapping the (timerdev, cc-channel) tuple to logical lines as done in the ADC and DAC interfaces? This would IMHO not be feasible here, as the programmed resolution and freq is the same for all channels of a timer device.

If you mean simple remapping of MCU timer cc-channels to logical RIOT channels (e.g. tim4-ch3 -> PWM_0-CH0), this is possible out of the box.

**Comment by DipSwitch** [comment]
The last one so its ok then :)

**Comment by haukepetersen** [comment]
yep.

**Comment by A-Paul** [comment]
@haukepetersen should this be closed in favor of #4638?

**Comment by haukepetersen** [comment]
yes.

**Pull request closed without merging**